File: LowestCommonAncestorofaBinarySearchTree.java
// Time Complexity : O(h) height of the binary tree
// Space Complexity : O(h)

/*
 * here we find the value of p and q in the binary tree if
 * the root is greater or smaller we move according to it. 
 * if the values are on the different ends of the tree we return root.
 */

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

 class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if(root.val > p.val && root.val > q.val) return lowestCommonAncestor(root.left, p, q);
        else if(root.val < p.val && root.val < q.val) return lowestCommonAncestor(root.right, p, q);
        else return root;
    }
}

File: lowest-common-ancestor-of-a-binary-search-tree.cpp
//Time - O(N)
//Space - O(1)
class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        if(root == NULL) return NULL; 
        
        while(root){
            if(p->val < root->val && q->val < root->val){
                root = root->left;
            }else if(p->val > root->val && q->val > root->val){
                root = root->right;
            }else{
                return root;
            }
        }
        
        return NULL;
        
    }
    
};